core/src/main/java/io/grpc/internal/BinaryLog.java

```diff
@@ -0,0 +1,208 @@
+          throw new IllegalArgumentException("Bad input: " + configuration);
+        }
+        String methodOrSvc = configMatcher.group(1);
+        String binlogOptionStr = configMatcher.group(2);
+        BinaryLog binLog = createBinaryLog(binlogOptionStr);
+        if (methodOrSvc.equals("*")) {
+          if (globalLog != null) {
+            throw new IllegalArgumentException("Duplicate log config for: *");
+          }
+          globalLog = binLog;
+          logger.info("Global binlog: " + globalLog);
+        } else if (isServiceGlob(methodOrSvc)) {
+          String service = MethodDescriptor.extractFullServiceName(methodOrSvc);
+          if (perServiceLogs.containsKey(service)) {
+            throw new IllegalArgumentException("Duplicate log config for service: " + methodOrSvc);
+          }
+          perServiceLogs.put(service, binLog);
+          logger.info(String.format("Service binlog: service=%s log=%s", service, binLog));
+        } else {
+          // assume fully qualified method name
+          if (perMethodLogs.containsKey(methodOrSvc)) {
+            throw new IllegalArgumentException("Duplicate log config for method: " + methodOrSvc);
+          }
+          perMethodLogs.put(methodOrSvc, binLog);
+          logger.info(String.format("Method binlog: method=%s log=%s", methodOrSvc, binLog));
+        }
+      }
+      enabled = true;
+      this.globalLog = globalLog == null ? NOOP_LOG : globalLog;
+    }
+
+    /**
+     * Accepts a full method name and returns the log that should be used.
+     */
+    public BinaryLog getLog(String fullMethodName) {
+      if (!enabled) {
+        return NOOP_LOG;
+      }
+      BinaryLog methodLog = perMethodLogs.get(fullMethodName);
+      if (methodLog != null) {
+        return methodLog;
+      }
+      BinaryLog serviceLog = perServiceLogs.get(
+          MethodDescriptor.extractFullServiceName(fullMethodName));
+      if (serviceLog != null) {
+        return serviceLog;
+      }
+      return globalLog;
+    }
+
+    /**
+     * Returns a binlog with the correct header and message limits. The input should be a string
+     * that is in one of these forms:
+     *
+     * <p>{@code {h(:\d+)?}, {m(:\d+)?}, {h(:\d+)?,m(:\d+)?}}
+     *
+     * <p>If the {@code logConfig} is null, the returned binlog will have a limit of
+     * Integer.MAX_VALUE.
+     */
+    @VisibleForTesting
+    static BinaryLog createBinaryLog(@Nullable String logConfig) {
+      if (logConfig == null) {
+        return new BinaryLog(Integer.MAX_VALUE, Integer.MAX_VALUE);
+      }
+      Matcher headerMatcher;
+      Matcher msgMatcher;
+      Matcher bothMatcher;
+      final int maxHeaderBytes;
+      final int maxMsgBytes;
+      if ((headerMatcher = headerRe.matcher(logConfig)).matches()) {
+        maxMsgBytes = 0;
+        String maxHeaderStr = headerMatcher.group(1);
+        maxHeaderBytes = maxHeaderStr != null ? Integer.parseInt(maxHeaderStr) : Integer.MAX_VALUE;
```

Comment:
Also handle the parse exception cleanly. (Have a helper that returns -1, and then check at the end? Or catch outside this method.)

Reply:
Originally I didn't think it was too wrong here to just let the constructor fail in static initialization and prevent app startup if the config string is malformed, since this naturally surfaces the error to the user. Same for the ambiguous case of duplicate log patterns. I can change it to be more graceful though.

Reply:
I'm fine with failing the user. But failing static initialization yields _weird_ results. Far too much confusion.